google/cloud/internal/future_then_async.h

```diff
@@ -0,0 +1,89 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_FUTURE_THEN_ASYNC_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_FUTURE_THEN_ASYNC_H
+/**
+ * @file
+ *
+ * Define the `then_async` template function.
+ *
+ * This function allows for generic asynchronous continuations using
+ * google::cloud::future. In particular, it allows chaining of async
+ * operations without increasing the nesting level.
+ * 
+ * Usage of this utility would look like the below, which is a series of
+ * asynchronous operations written in an almost linear style:
+ * 
+ * future<std::string> f1 = ...
+ * future<int> f2 = then_async(std::move(f1), [](std::string value) -> future<int> {
+ *     return StartAsyncOp(std::move(value));
+ * });
+ * future<Status> f3 = then_async(std::move(f2), [](int value) -> future<Status> {
+ *     return FinishAsyncOp(value);
+ * });
+ * return f3;
+ */
+
+#include "google/cloud/future.h"
+#include "google/cloud/version.h"
+#include "google/cloud/internal/invoke_result.h"
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace internal {
+
+template <class FutureT>
+struct future_value_type;
+template <class T>
+struct future_value_type<future<T>> {
+    using type = T;
+};
+template <class T>
+using future_value_type_t = typename future_value_type<T>::type;
+
+void link_future_and_promise(future<void> fut, promise<void> p) {
+    fut.then([p]{ p.set_value(); });
+}
+
+template <class T>
+void link_future_and_promise(future<T> fut, promise<T> p) {
+    fut.then([p](T val){ p.set_value(std::move(val)); });
```

Comment:
`promise<T>` is move-only, and this is C++ 11, you need to implement a helper class to move-capture.